src/inout.cpp

```diff
@@ -0,0 +1,61 @@
+#include "inout.h"
+#include "mongoose.h"
+
+namespace graft
+{
+std::unordered_map<std::string, std::string> OutHttp::uri_substitutions;
+
+//following #define depends on hm such as const http_message&
+#define SET_STR_FLD(fld) \
+    { \
+        if(hm.fld.len == 0) \
+        { \
+            fld.clear(); \
+        } \
+        else \
+        { \
+            int off = hm.fld.p - hm.message.p; \
+            assert(0<hm.fld.len && 0<=off && off+hm.fld.len<=hm.message.len); \
+            fld = std::string(hm.fld.p, hm.fld.len); \
+        } \
+    }
+#define SET_ALL_FLD() \
+    { \
+        SET_STR_FLD(body); \
+        SET_STR_FLD(method); \
+        SET_STR_FLD(uri); \
+        SET_STR_FLD(proto); \
+        resp_code = hm.resp_code; \
+        SET_STR_FLD(resp_status_msg); \
+        SET_STR_FLD(query_string); \
+    }
+InOutHttpBase& InOutHttpBase::operator = (const http_message& hm)
```

Comment:
Generally speaking, I'm not sure I like the idea of assigning objects of different and unrelated types to each other

Reply:
If we have constructor InOutHttpBase(const http_message& hm), it is natural to have assignment operator from the same type. Is not?

Reply:
Good example, there are both 
basic_string& operator=( const CharT* s );
basic_string( const CharT* s, const Allocator& alloc = Allocator() );
What don't I understand? 

Maybe you want to say that they are not relates as string and char*. But
    using comp = std::complex<int>;
    comp a(1,2);
    //but
    comp c(1);
    c = 2;

Reply:
OK, convinced :)